Added unit testing section to the client README (#7890)
galaxyproject/galaxy: client/README.md

On build, the compiled css bundle is served at `/static/style/base.css`.

As mentioned above, `make client` will rebuild styles, as a part of the webpack build.  For iterative development, "Watch Mode" rebuilds as described above do include style changes.

# Client-Side Unit Testing

Galaxy's client is undergoing an extensive refactoring and modernizing process.

> As part of initiatve, we would like to request that all new client-side code

Reviewer: ```suggestion
As part of this initiative, we would like to request that all new client-side code
```

#### Run qunit legacy tests

    npm run test-qunit

## Writing a test file

The karma configs are setup to look at any file ending in "*.test.js". Please
place your test files inside client/galaxy/scripts folders right next to
whatever files that they are testing.

```javascript
// yourtestfile.test.js

import { things } from "./yourtestfile.js";

describe("some module you wrote", () => {

    let transientVariables, serviceInstances, testData;

    beforeEach(() => {
        // setup your code (if necessary)
    })

    afterEach(() => {
        // teardown your code (so it doesn't ruin the next test)
    })

    it("should do something or other", () => {
        assert(somethingThatShouldTrue, "explain what went wrong");
        expect(something).to.eq(true);
    })
})
```

## Testing ~~Suggestions~~ Obligations

### Clearly document the intent of your test

Please remember that these tests are not *for* you. They're for the people who
come after you. It will be a lot easier to modify, repair and upgrade your code
if they can figure out what you were originally hoping to accomplish.

> Add a couple comments. Use variable names that mean something. Use the assertion

Reviewer: ```suggestion
Add a couple of comments. Use variable names that mean something. Use the assertion
```

failure description parameters when possible. I'm not sure what your mom told

> you, but your code is almost definitely not self-documenting.

Reviewer 2: I think we should just drop the your mom statement, it'd still convey the same message.

Author: Removed! I didn't trust your mom anyway.

broken up into easily testable chunks.

## Specific test scenarios & examples

[Mocking an imported dependency](https://github.com/galaxyproject/galaxy/blob/dev/client/galaxy/scripts/components/Tags/tagService.test.js)

[Testing async operations](https://github.com/galaxyproject/galaxy/blob/dev/client/galaxy/scripts/components/Tags/tagService.test.js)

[Testing a Vue component for expected rendering output](https://github.com/galaxyproject/galaxy/blob/dev/client/galaxy/scripts/components/Tags/StatelessTags.test.js)

[Firing an event against a shallow mounted vue component](https://github.com/galaxyproject/galaxy/blob/dev/client/galaxy/scripts/components/Tags/StatelessTags.test.js)

## The dirty secret about testing

It's good to have tests, but testing code isn't really about testing at all,
it's actually about software design. 

Maintainable software is testable software. If you can run a unit test on your
code, then that means you must have necessarily separated your code into
testable units and you will have almost definitely written better, more modular,
more easily manipulated code, and nobody will ever contemplate using git-blame

> figure out what went wrong.

Reviewer: ```suggestion
to figure out what went wrong.
```

> to hunt you down.

Reviewer 2: ```suggestion
figure out what went wrong.
```
